server.go

```diff
@@ -666,8 +692,15 @@ func (srv *Server) readTCP(conn net.Conn, timeout time.Duration) ([]byte, error)
 		return nil, err
 	}
 
-	m := make([]byte, length)
+	m := srv.tcpPool.Get().([]byte)
+	if int(length) > len(m) {
+		srv.tcpPool.Put(m)
+		return nil, ErrBuf
```

Comment:
wouldn't it be better to allocate new buffer here instead of dropping the query and eventually closing the connection?

Reply:
Tbh, I am not sure about that. On one hand, you're right, we may want to process ALL queries. On the other hand, we would need to allocate a 64kb buffer that will end up in the `sync.Pool` which is an undesirable outcome.

Also, despite that DNS queries aren't limited by 512 bytes by the RFC, I hardly can imagine a situation when a DNS query is bigger than that.

Reply:
> we would need to allocate a 64kb buffer

you can allocate the buffer of the size exactly needed for reading this particular message, since you already know the message size.

> that will end up in the sync.Pool which is an undesirable outcome.

you don't need to put the big buffers to the pool, i.e. check the buffer capacity before putting back.